### PR TITLE
183 adjust field

### LINKED ISSUE #183: Adjust field implementation
As part of the CETACEO project, one of the objectives requested by Airbus is to interpolate the aerodynamic mesh Cp values onto a structural mesh while preserving one or more global quantities, such as CL or CD. Our proposed approach is to use one of the models available in `pyLOM.NN` (MLP, KANN, GNN) to predict Cp over each element of the structural mesh, and then apply an adjustment that minimally modifies the prediction so that the global CL and CD match those from the aerodynamic mesh.

This adjustment is formulated as a constrained optimization problem where the unknowns are the entire pressure field on the aerodynamic surface. I tested several methods to solve this problem. Due to the high number of variables (depending on the mesh density, currently ranging from 20,000 to 200,000 elements), classical optimization methods are too slow and memory-intensive, especially when building the Hessian matrix.
The best results were obtained using PyTorch’s Adam optimizer, which applies gradient descent to find the optimal solution.

Since this process needs to be part of pyLOM, I considered implementing a new method `adjust_field()` within the `pyLOM.Dataset` class. This method would allow adjusting a field according to a user-defined objective function. I’ve tried to design it in the most generic way possible so that it’s not limited to this specific use case, but can be applied to adjust any field based on any differentiable objective function.

Please let me know your thoughts. If you think this is too specific to include in `pyLOM.Dataset`, an alternative would be to implement a similar method in `pyLOM.NN.Dataset`. However, given how each dataset is structured, it seemed more natural and straightforward to implement it in `pyLOM.Dataset`.

And let me know if anything I’ve explained is unclear or needs further detail. 

```python 
def adjust_field(
        self,
        fieldname,
        obj_func,
        get_opt_param_func,
        constr_func=None,
        optimizer_class=torch.optim.Adam, 
        schduler_class=torch.optim.lr_scheduler.StepLR,
        opt_config=None,
        disp_progress=(False, 0),
        **kwargs
    ):
        default_config = {
            'niter': 1000,
            'lr': 1e-2,
            'lr_step_size': 1,
            'lr_gamma': 0.999,
            'penalty_factor': 1e5,
            'tolerance': 1e-9,
            'patience': 10,
        }

        if opt_config is None:
            opt_config = {}
        config = {**default_config, **opt_config}

        field = self[fieldname]
        field_mod = field.copy()
        field_losses = []

        def closure(colTensor, colTensor0, opt_vars):
            optimizer.zero_grad()
            obj_loss = obj_func(colTensor, colTensor0, **opt_vars)
            if constr_func is not None:
                penalty = constr_func(colTensor, **opt_vars)
                total_loss = obj_loss + config['penalty_factor'] * penalty
                total_loss.backward()
                return total_loss, obj_loss, penalty
            else:
                total_loss = obj_loss
                total_loss.backward()
                return total_loss, obj_loss, torch.tensor(0.0)

        for i, col in enumerate(field.T):
            colTensor = torch.tensor(col, requires_grad=True)
            colTensor0 = colTensor.clone().detach()
            opt_vars = get_opt_param_func(self, i, **kwargs)

            optimizer = optimizer_class([colTensor], lr=config['lr'])
            scheduler = schduler_class(optimizer, step_size=config['lr_step_size'], gamma=config['lr_gamma'])
            losses = []

            if disp_progress[0]:
                print(f"\nCase {i}:")

            prev_loss = float('inf')
            n_improvement = 0

            for epoch in range(config['niter']):
                total_loss, obj_loss, penalty = optimizer.step(lambda: closure(colTensor, colTensor0, opt_vars))
                scheduler.step()
                losses.append([obj_loss.item(), penalty.item(), total_loss.item()])

                if disp_progress[0] and (epoch % disp_progress[1] == 0):
                    print(f"Epoch {epoch:4}: Total Loss = {total_loss.item():.2e}, Objective = {obj_loss.item():.2e}, Penalty = {penalty.item():.2e}")

                loss_diff = abs(prev_loss - total_loss.item())
                prev_loss = total_loss.item()
                if loss_diff < config['tolerance']:
                    n_improvement += 1
                else:
                    n_improvement = 0
                if n_improvement >= config['patience']:
                    print(f"Early stopping at epoch {epoch}, no significant improvement.")
                    break
                if epoch >= config['niter'] - 1:
                    print(f"Reached maximum number of epochs ({config['niter']}). Stopping.")
                    break

            field_mod[:, i] = colTensor.detach().numpy()
            field_losses.append(losses)

        ndim = self.info(fieldname)['ndim']
        self.add_field(varname=fieldname + 'Adjusted', ndim=ndim, var=field_mod)
        return field_losses
```

### PR DESCRIPTION
**Pull request by fermingut**
Closes #183

### CODE REVIEW

File: pyLOM/NN/aerodynamics.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            CoefDragPressure = ForcePressure[2]*torch.sin(alpha) + ForcePressure[0]*torch.cos(alpha)
+        else:
+            raise ValueError("CoefPressure must have shape (n,) or (n,1).")
+    
+    # Handle CoefSkinFriction
+    if CoefSkinFriction is not None:
+        if CoefSkinFriction.ndimension() == 1:
+            CoefSkinFriction = CoefSkinFriction.view(-1, 3)
+        elif CoefSkinFriction.shape[1] != 3:
+            raise ValueError("CoefSkinFriction must have shape (n,3).")
```
Comment (ArnauMiro) [inline pyLOM/NN/aerodynamics.py:53]: Use the `raiseError` method of `pyLOM`

File: pyLOM/NN/aerodynamics.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        
+        Si = torch.sqrt(normals[:, 0]**2 + normals[:, 1]**2 + normals[:, 2]**2).view(-1, 1)
+        ForceFriction = CoefSkinFriction * Si
+        CoefLiftFriction = torch.sum(ForceFriction[:, 2] * torch.cos(alpha)) - torch.sum(ForceFriction[:, 0] * torch.sin(alpha))
+        CoefDragFriction = torch.sum(ForceFriction[:, 2] * torch.sin(alpha)) + torch.sum(ForceFriction[:, 0] * torch.cos(alpha))
+    
+    if CoefPressure is None and CoefSkinFriction is None:
+        raise ValueError("At least one of CoefPressure or CoefSkinFriction must be provided.")
```
Comment (ArnauMiro) [inline pyLOM/NN/aerodynamics.py:61]: Use the `raiseError` method of `pyLOM`

File: pyLOM/NN/interpolator.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        ref_values = {}
+        penalty_args = {}
+
+        for original_name, (source_type, func_arg_name) in mapping.items():
+            if source_type == 'get_variable':
+                val = dataset.get_variable(original_name)[i]
+            elif source_type == 'field':
+                val = dataset[original_name]
+            else:
+                raise ValueError(f"Unknown source type '{source_type}' for variable '{original_name}'")
```
Comment (ArnauMiro) [inline pyLOM/NN/interpolator.py:100]: Use the `raiseError` method of `pyLOM`

File: pyLOM/NN/interpolator.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            colTensor = torch.tensor(col, requires_grad=True)
+            colTensor0 = colTensor.clone().detach()
+            opt_vars = get_opt_param_func(self.dataset, i, **kwargs)
+
+            optimizer = optimizer_class([colTensor], lr=config['lr'])
+            scheduler = schduler_class(optimizer, step_size=config['lr_step_size'], gamma=config['lr_gamma'])
+            losses = []
+
+            if disp_progress[0]:
+                print(f"\nCase {i}:")
```
Comment (ArnauMiro) [inline pyLOM/NN/interpolator.py:191]: Use the `pprint` method of `pyLOM`

File: pyLOM/NN/interpolator.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            prev_loss = float('inf')
+            n_improvement = 0
+
+            for epoch in range(config['niter']):
+                total_loss, obj_loss, penalty = optimizer.step(lambda: closure(colTensor, colTensor0, opt_vars))
+                scheduler.step()
+                losses.append([obj_loss.item(), penalty.item(), total_loss.item()])
+
+                if disp_progress[0] and (epoch % disp_progress[1] == 0):
+                    print(f"Epoch {epoch:4}: Total Loss = {total_loss.item():.2e}, Objective = {obj_loss.item():.2e}, Penalty = {penalty.item():.2e}")
```
Comment (ArnauMiro) [inline pyLOM/NN/interpolator.py:202]: Use the `pprint` method of `pyLOM`

File: pyLOM/NN/interpolator.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                prev_loss = total_loss.item()
+                if loss_diff < config['tolerance']:
+                    n_improvement += 1
+                else:
+                    n_improvement = 0
+                if n_improvement >= config['patience']:
+                    print(f"Early stopping at epoch {epoch}, no significant improvement.")
+                    break
+                if epoch >= config['niter'] - 1:
+                    print(f"Reached maximum number of epochs ({config['niter']}). Stopping.")
```
Comment (ArnauMiro) [inline pyLOM/NN/interpolator.py:214]: Use the `pprint` method of `pyLOM`
